tools/analyzers/recursivelock/analyzer.go

```diff
@@ -0,0 +1,455 @@
+		(*ast.FuncLit)(nil),
+		(*ast.File)(nil),
+		(*ast.ReturnStmt)(nil),
+	}
+
+	var keepTrackOf tracker
+	inspect.Preorder(nodeFilter, func(node ast.Node) {
+		if keepTrackOf.funcLitEnd.IsValid() && node.Pos() <= keepTrackOf.funcLitEnd {
+			return
+		} else {
+			keepTrackOf.funcLitEnd = token.NoPos
+		}
+		if keepTrackOf.deferEnd.IsValid() && node.Pos() > keepTrackOf.deferEnd {
+			keepTrackOf.deferEnd = token.NoPos
+		} else if keepTrackOf.deferEnd.IsValid() {
+			return
+		}
+		if keepTrackOf.retEnd.IsValid() && node.Pos() > keepTrackOf.retEnd {
+			keepTrackOf.retEnd = token.NoPos
+			keepTrackOf.incFRU()
+		}
+
+		switch stmt := node.(type) {
+		case *ast.CallExpr:
+			call := getCallInfo(pass.TypesInfo, stmt)
+			if call == nil {
+				break
+			}
+			name := call.name
+			selMap := mapSelTypes(stmt, pass)
+			if selMap == nil {
+				break
+			}
+			if keepTrackOf.rLockSelector != nil {
+				if keepTrackOf.foundRLock > 0 {
+					if keepTrackOf.rLockSelector.isEqual(selMap, 0) {
+						pass.Reportf(
+							node.Pos(),
+							fmt.Sprintf(
+								"%v",
+								errNestedRLock,
+							),
+						)
+					} else {
+						if stack := hasNestedRLock(keepTrackOf.rLockSelector, selMap, call, inspect, pass, make(map[string]bool)); stack != "" {
+							pass.Reportf(
+								node.Pos(),
+								fmt.Sprintf(
+									"%v\n%v",
+									errNestedRLock,
+									stack,
+								),
+							)
+						}
+					}
+				}
+				if name == "RUnlock" && keepTrackOf.rLockSelector.isEqual(selMap, 1) {
+					keepTrackOf.deincFRU()
+				}
+			} else if name == "RLock" && keepTrackOf.foundRLock == 0 {
+				keepTrackOf.rLockSelector = selMap
+				keepTrackOf.incFRU()
+			}
+
+		case *ast.File:
+			keepTrackOf = tracker{}
+
+		case *ast.FuncDecl:
+			keepTrackOf = tracker{}
+			keepTrackOf.funcEnd = stmt.End()
+
+		case *ast.FuncLit:
+			if keepTrackOf.funcLitEnd == token.NoPos {
+				keepTrackOf.funcLitEnd = stmt.End()
+			}
+
+		case *ast.DeferStmt:
+			call := getCallInfo(pass.TypesInfo, stmt.Call)
+			if keepTrackOf.deferEnd == token.NoPos {
+				keepTrackOf.deferEnd = stmt.End()
+			}
+			if call != nil && call.name == "RUnlock" {
+				keepTrackOf.deferredRUnlock = true
+			}
+
+		case *ast.ReturnStmt:
+			if keepTrackOf.deferredRUnlock && keepTrackOf.retEnd == token.NoPos {
+				keepTrackOf.deincFRU()
+				keepTrackOf.retEnd = stmt.End()
+			}
+		}
+	})
+	return nil, nil
+}
+
+type tracker struct {
+	funcEnd         token.Pos
+	retEnd          token.Pos
+	deferEnd        token.Pos
+	funcLitEnd      token.Pos
+	deferredRUnlock bool
+	foundRLock      int
+	rLockSelector   *selIdentList
+}
+
+func (t tracker) String() string {
+	return fmt.Sprintf("funcEnd:%v\nretEnd:%v\ndeferEnd:%v\ndeferredRU:%v\nfoundRLock:%v\n", t.funcEnd, t.retEnd, t.deferEnd, t.deferredRUnlock, t.foundRLock)
+}
+
+func (t *tracker) deincFRU() {
+	if t.foundRLock > 0 {
+		t.foundRLock -= 1
+	}
+}
+func (t *tracker) incFRU() {
+	t.foundRLock += 1
+}
+
+type selIdentNode struct {
```

Comment:
Maybe some comments on why `selIdentNode` and `selIdentList` is needed and what they are used for. Also some comments on their respective methods and what they do would be helpful too.